clippy_lints/src/loops/while_let_on_iterator.rs

```diff
@@ -1,170 +1,353 @@
-use super::utils::{LoopNestVisitor, Nesting};
 use super::WHILE_LET_ON_ITERATOR;
 use clippy_utils::diagnostics::span_lint_and_sugg;
 use clippy_utils::source::snippet_with_applicability;
-use clippy_utils::ty::implements_trait;
-use clippy_utils::usage::mutated_variables;
-use clippy_utils::{
-    get_enclosing_block, is_refutable, is_trait_method, last_path_segment, path_to_local, path_to_local_id,
-};
-use if_chain::if_chain;
+use clippy_utils::{get_enclosing_loop, is_refutable, is_trait_method, match_def_path, paths, visitors::is_res_used};
 use rustc_errors::Applicability;
-use rustc_hir::intravisit::{walk_block, walk_expr, NestedVisitorMap, Visitor};
-use rustc_hir::{Expr, ExprKind, HirId, MatchSource, Node, PatKind};
+use rustc_hir::intravisit::{walk_expr, ErasedMap, NestedVisitorMap, Visitor};
+use rustc_hir::{def::Res, Expr, ExprKind, HirId, Local, MatchSource, Node, PatKind, QPath, UnOp};
 use rustc_lint::LateContext;
-use rustc_middle::hir::map::Map;
-use rustc_span::symbol::sym;
+use rustc_span::{symbol::sym, Span, Symbol};
 
 pub(super) fn check(cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) {
-    if let ExprKind::Match(match_expr, arms, MatchSource::WhileLetDesugar) = expr.kind {
-        let pat = &arms[0].pat.kind;
-        if let (&PatKind::TupleStruct(ref qpath, pat_args, _), &ExprKind::MethodCall(method_path, _, method_args, _)) =
-            (pat, &match_expr.kind)
-        {
-            let iter_expr = &method_args[0];
-
-            // Don't lint when the iterator is recreated on every iteration
-            if_chain! {
-                if let ExprKind::MethodCall(..) | ExprKind::Call(..) = iter_expr.kind;
-                if let Some(iter_def_id) = cx.tcx.get_diagnostic_item(sym::Iterator);
-                if implements_trait(cx, cx.typeck_results().expr_ty(iter_expr), iter_def_id, &[]);
-                then {
-                    return;
-                }
-            }
+    if let ExprKind::Match(scrutinee_expr, [arm, _], MatchSource::WhileLetDesugar) = expr.kind {
```

Comment:
Personally I would prefer a name like `match_expr` instead of `scrutinee_expr`. _Scrutinee_ is the correct therm, but I had to look it up while `match_expr` would ensure that everybody understands it right away.

Reply:
`match_expr` could be confused for the actual `match` expression. Probably not in this case, but it is ambiguous to name it like that.

Reply:
Fair point :upside_down_face:

Reply:
Ah, you've combined the two `match` statements into a single `if_chain!`. I thought more about implementing them as two chains. This looks good at a first glance, so what ever works ^^. I'll try to review all changes and finish the review this week, however it'll probably be pushed into the weekend :upside_down_face:

Reply:
There isn't really any reason to have two chains one after another. I did return from the chain rather than put everything into the `then` block so rustfmt will touch it.

Reply:
If you have an unambiguous name other than scrutinee I can switch to it. Personally I think we should switch to using it everywhere so everyone can learn it.

Reply:
I think it would be good if we use the correct therms for the syntax, the only problem is that it can increase the learning curve for newcomers to the project. That was also the main reason for the suggestion your arguments are compelling, and I was happy to learn a new therm. So, you can resolve this thread as well, at least from my side :)

Reply:
Sounds good.